src/main/java/uk/gov/companieshouse/overseasentitiesapi/interceptor/UserAuthenticationInterceptor.java

```diff
@@ -74,4 +110,52 @@ public boolean preHandle(@NonNull HttpServletRequest request, @NonNull HttpServl
     protected Optional<TokenPermissions> getTokenPermissions(HttpServletRequest request) {
         return AuthorisationUtil.getTokenPermissions(request);
     }
+
+    protected boolean isRequestRoeUpdate(HttpServletRequest request) {
+        final Map<String, List<String>> privileges = getERICTokenPermissions(request);
+        if (privileges.containsKey(COMPANY_NUMBER_KEY)) {
+            companyNumberInScope = privileges.get(COMPANY_NUMBER_KEY).get(0);
+            return StringUtils.isNotEmpty(companyNumberInScope) && companyNumberInScope.startsWith("OE");
+            }
+        return false;
+    }
+
+    protected boolean doCompanyNumbersMatch(HttpServletRequest request, HttpServletResponse response, String transactionId) {
+        String companyNumberInTransaction = getCompanyNumberInTransaction(request, response, transactionId);
+        if (StringUtils.isNotBlank(companyNumberInTransaction)) {
+            return companyNumberInTransaction.equalsIgnoreCase(companyNumberInScope);
+        }
+        return false;
+    }
+
+    private Map<String, List<String>> getERICTokenPermissions(HttpServletRequest request) {
+        String tokenPermissionsHeader = request.getHeader(ERIC_AUTHORISED_TOKEN_PERMISSIONS);
+        Map<String, List<String>> permissions = new HashMap<>();
+        if (tokenPermissionsHeader != null) {
+            for (String pair : tokenPermissionsHeader.split(" ")) {
+                String[] parts = pair.split("=");
+                permissions.put(parts[0], Arrays.asList(parts[1].split(",")));
+            }
+        }
+        return permissions;
+    }
+
+    private String getCompanyNumberInTransaction(HttpServletRequest request, HttpServletResponse response, String transactionId) {
+        String passthroughHeader = request.getHeader(ApiSdkManager.getEricPassthroughTokenHeader());
+        String reqId = request.getHeader(ERIC_REQUEST_ID_KEY);
+        var logMap = new HashMap<String, Object>();
+        logMap.put(TRANSACTION_ID_KEY, transactionId);
+        try {
+            final var transaction = transactionService.getTransaction(transactionId, passthroughHeader, reqId);
```

Comment:
I have some concerns about calling out to an external service within a (user auth) interceptor. Possibly lengthy delay or failure. Alternatively we could cache the company number and do the checks in the request handling.

Reply:
In theory the call is fine - we make it in another interceptor anyway - but we could possibly avoid doing it and just reference the transaction that's (hopefully) already retrieved by doing:

        final var transaction = (Transaction) request.getAttribute(TRANSACTION_KEY);

See the `AbstractTransactionStatusInterceptor` class which does this. However, as well as unit-testing this works, you'll need to test with the app, just due to the order in which the interceptors are invoked at run-time.

Reply:
To address this we get the transaction only as a last final step in the verification process only if all other checks have passed.